### ISSUE TITLE
Should bleach sanitize certain python escape characters?

### ISSUE THREAD

**Issue by jvanasco** [mozilla/bleach#298]
doing an audit and updating some tests, I decided to put bleach through some of python's escape characters.

* https://docs.python.org/2/reference/lexical_analysis.html#string-literals
* https://docs.python.org/3.6/reference/lexical_analysis.html#string-and-bytes-literals

These were largely left as-is, and probably should be sanitized:

* bell
* vertical tab
* backspace
* formfeed

there is also the `\N{name}` (unicode decode swap), and `\ooo` (octal), and `\xhh` (hex) ways to drop in some of these chars as well.

# questionable 

Bell `\a`

	>>> print(bleach.clean("a\ac"))
	ac
	>>> print(bleach.clean("a\ac")).__repr__()
	u'a\x07c'

Vertical Tab `\v`

	>>> print(bleach.clean("a\vc"))
	a
	 c
	>>> print(bleach.clean("a\vc")).__repr__()
	u'a\x0bc'

Backspace `\b`

	>>> print(bleach.clean("a\bc"))
	c
	>>> print(bleach.clean("a\bc")).__repr__()
	u'a\x08c'

Formfeed `\f`

	>>> print(bleach.clean("a\fc"))
	c
	>>> print(bleach.clean("a\fc")).__repr__()
	u'a\x0cc'

# ok

Carriage Return	`\r`

	>>> print(bleach.clean("a\rc"))
	a
	c
	>>> print(bleach.clean("a\rc")).__repr__()
	u'a\nc'

Line Feed `\n`
	>>> print(bleach.clean("a\nc"))
	a
	c
	>>> print(bleach.clean("a\nc")).__repr__()
	u'a\nc'

Tab `\n`
	>>> print(bleach.clean("a\tc"))
	a	c
	>>> print(bleach.clean("a\tc")).__repr__()
	u'a\tc'

**Comment by willkg** [comment]
Do any of these create unsafe HTML scenarios? If so, please write up a security bug. If not, then I don't think this is something Bleach should be doing.

**Comment by willkg** [comment]
I spent a bunch of time looking into this. A web browser (Firefox, links, etc) will show glyphs for characters covered by this issue. There's no way anyone would look at that and want to copy/paste it.

However, printing `bleach.clean()` output to a terminal like this will interpret these characters causing the above problem because that's not a web browser rendering HTML--it's just printing a string to a terminal.

Regardless, it seems helpful to sanitize characters in Characters tokens by converting them to the relevant character entities.

Using the above example, after this change we'd get something like this:

```
>>> print(bleach.clean("a\ac"))
a&#x07;c
```

I think that should work fine in web browsers, too. I'll verify that when I get there.

I'll try to get this done in the next few days and then do a Bleach 2.1 release.

**Comment by willkg** [comment]
I'm going to fix this for Bleach 2.1, but not for older versions of Bleach. That's going to be a problem for anyone who hasn't updated to Bleach 2, yet.

For those people, I suggest doing something like the following:

```
def remove_invisible_characters(text):
    for c in ('\a', '\b', '\f', '\v'):
        text = text.replace(c, '')
    return text
```

Then wrap `bleach.clean()` calls with that like this:

```
text = remove_invisible_characters(bleach.clean(dirty_text))
```

**Comment by jvanasco** [comment]
Thanks for handling this, Will!

For the benefit of others- Will & I had a side-chat on this on this in the security tracker.  Some, not all, browsers will display glyphs for the ASCII control characters.  The newest version of Firefox and most versions of text browsers will display glyphs, but Chrome/Safari and older firefox versions will show the empty space.

**Comment by Rogalek** [comment]
Just to clarify, so `\n` character is good to go and it is not cleaned?

**Comment by jvanasco** [comment]
@Rogalek yes. \n only affects display in html source or within ‘pre’ tags. You may want to run a regex to replace all and duplicate white space with a single space.
